PhysicsTools/PatAlgos/plugins/CandidateSummaryTable.cc

```diff
@@ -12,65 +12,89 @@
 */
 
 
-#include "FWCore/Framework/interface/EDAnalyzer.h"
+#include "FWCore/Framework/interface/stream/EDAnalyzer.h"
 #include "FWCore/Framework/interface/Event.h"
 #include "FWCore/Framework/interface/ConsumesCollector.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/Utilities/interface/InputTag.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include <iomanip>
+#include <atomic>
 
 #include "DataFormats/Common/interface/View.h"
 #include "DataFormats/Candidate/interface/Candidate.h"
 
-namespace pat {
-  class CandidateSummaryTable : public edm::EDAnalyzer {
-    public:
-      explicit CandidateSummaryTable(const edm::ParameterSet & iConfig);
-      ~CandidateSummaryTable();
-
-      virtual void analyze(const edm::Event & iEvent, const edm::EventSetup& iSetup) override;
-      virtual void endJob() override;
-
-    private:
-      struct Record {
-        edm::InputTag src;
-        edm::EDGetTokenT<edm::View<reco::Candidate> > srcToken;
-        size_t present, empty, min, max, total;
-        Record(edm::InputTag tag, edm::ConsumesCollector && iC) : src(tag), srcToken(iC.consumes<edm::View<reco::Candidate> >(tag)), present(0), empty(0), min(0), max(0), total(0) {}
-
-        void update(const edm::View<reco::Candidate> &items) {
-            present++;
-            size_t size = items.size();
-            if (size == 0) {
-                empty++;
-            } else  {
-                if (min > size) min = size;
-                if (max < size) max = size;
-            }
-            total += size;
-        }
-      };
-      std::vector<Record> collections_;
-      size_t totalEvents_;
-      bool perEvent_, perJob_;
-      std::string self_, logName_;
-      bool dumpItems_;
+namespace pathelpers {
+  struct Record {
+    const edm::InputTag src;
+    mutable std::atomic<size_t> present, empty, min, max, total;
+    Record(const edm::InputTag& tag) : src(tag), present(0), empty(0), min(0), max(0), total(0) {}
+    Record(const Record& other) : src(other.src), 
+                                  present(other.present.load()), 
+                                  empty(other.empty.load()), 
+                                  min(other.min.load()), 
+                                  max(other.max.load()), 
+                                  total(other.total.load()) {}
+    void update(const edm::View<reco::Candidate> &items) const {
+      present++;
+      const size_t size = items.size();
+      if (size == 0) {
+        empty++;
+      } else  {
+        auto previousMin = min.load();
+        while( previousMin > size and not min.compare_exchange_weak(previousMin, size) ) {}
+        auto previousMax = max.load();
+        while( previousMax > size and not max.compare_exchange_weak(previousMax, size) ) {}        
```

Comment:
Thanks for the catch! I'll fix in a moment.

(Sent from my Nexus 6)
On Aug 12, 2015 10:52 PM, "Carl Vuosalo" <notifications@github.com> wrote:

> In PhysicsTools/PatAlgos/plugins/CandidateSummaryTable.cc
> <https://github.com/cms-sw/cmssw/pull/10620#discussion_r36912304>:
>
> > +    Record(const Record& other) : src(other.src),
> > +                                  present(other.present.load()),
> > +                                  empty(other.empty.load()),
> > +                                  min(other.min.load()),
> > +                                  max(other.max.load()),
> > +                                  total(other.total.load()) {}
> > +    void update(const edm::View<reco::Candidate> &items) const {
> > +      present++;
> > +      const size_t size = items.size();
> > +      if (size == 0) {
> > +        empty++;
> > +      } else  {
> > +        auto previousMin = min.load();
> > +        while( previousMin > size and not min.compare_exchange_weak(previousMin, size) ) {}
> > +        auto previousMax = max.load();
> > +        while( previousMax > size and not max.compare_exchange_weak(previousMax, size) ) {}
>
> To match the previous code, I think you want "previousMax < size" here.
> Otherwise, min and max will behave identically.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10620/files#r36912304>.
>